src/modules/lost/functions.c

```diff
@@ -698,70 +1055,128 @@ int lost_function(struct sip_msg *_m, char *_con, char *_uri, char *_name,
-	doc = xmlReadMemory(ret.s, ret.len, 0, 0,
-			XML_PARSE_NOBLANKS | XML_PARSE_NONET | XML_PARSE_NOCDATA);
-
-	if(doc == NULL) {
-		LM_ERR("invalid xml document: [%.*s]\n", ret.len, ret.s);
-		doc = xmlRecoverMemory(ret.s, ret.len);
-		if(doc == NULL) {
-			LM_ERR("xml document recovery failed on: [%.*s]\n", ret.len, ret.s);
-			goto err;
-		}
-
-		LM_DBG("xml document recovered\n");
-	}
-	root = xmlDocGetRootElement(doc);
-	if(root == NULL) {
-		LM_ERR("empty xml document: [%.*s]\n", ret.len, ret.s);
-		/* free memory */
-		lost_free_string(&ret);
-		goto err;
-	}
-	/* check the root element, shall be findServiceResponse, or errors */
-	if((!xmlStrcmp(root->name, (const xmlChar *)"findServiceResponse"))) {
-		/* get the uri element */
-		uri.s = lost_get_content(root, uri_element, &uri.len);
-		if(uri.len == 0) {
-			LM_ERR("uri element not found: [%.*s]\n", ret.len, ret.s);
-			/* free memory */
-			lost_free_string(&ret);
-			goto err;
+	/* at least parse one request */
+	redirect = 1;
+	while(redirect) {
+		fsrdata = lost_parse_findServiceResponse(ret);
+		if (lost_verbose == 1) {
+			lost_print_findServiceResponse(fsrdata);
 		}
-		LM_INFO("### LOST uri\t[%.*s]\n", uri.len, uri.s);
-		/* get the displayName element */
-		name.s = lost_get_content(root, name_element, &name.len);
-		if(name.len == 0) {
-			LM_ERR("displayName element not found: [%.*s]\n", ret.len, ret.s);
-			/* free memory */
-			lost_free_string(&ret);
-			goto err;
-		}
-		LM_INFO("### LOST din\t[%.*s]\n", name.len, name.s);
-	} else if((!xmlStrcmp(root->name, (const xmlChar *)"errors"))) {
-
-		LM_DBG("findService error response received\n");
-
-		/* get the error patterm */
-		err.s = lost_get_childname(root, errors_element, &err.len);
-		LM_DBG("findService error response: [%.*s]\n", err.len, err.s);
-		if(err.len == 0) {
-			LM_ERR("error pattern element not found: [%.*s]\n", ret.len, ret.s);
-			/* free memory */
-			lost_free_string(&ret);
-			goto err;
+		switch(fsrdata->category) {
+			case RESPONSE:
+				if(fsrdata->uri != NULL) {
+					/* get the first uri element */
+					if((tmp.s = fsrdata->uri->value) != NULL) {
+						tmp.len = strlen(fsrdata->uri->value);
+						uri.s = lost_copy_string(tmp, &uri.len);
+					}
+				} else {
+					LM_ERR("uri not found: [%.*s]\n", ret.len, ret.s);
+					goto err;
+				}
+				if(fsrdata->mapping != NULL) {
+					/* get the displayName element */
+					if((tmp.s = fsrdata->mapping->name->text) != NULL) {
+						tmp.len = strlen(fsrdata->mapping->name->text);
+						name.s = lost_copy_string(tmp, &name.len);
+					}
+				} else {
+					LM_ERR("name not found: [%.*s]\n", ret.len, ret.s);
+					goto err;
+				}
+				/* we are done */
+				redirect = 0;
+				break;
+			case ERROR:
+				/* get the errors element */
+				if(fsrdata->errors != NULL) {
+					if((tmp.s = fsrdata->errors->issue->type) != NULL) {
+						tmp.len = strlen(fsrdata->errors->issue->type);
+						err.s = lost_copy_string(tmp, &err.len);
```

Comment:
if you need a str anyway, you could use pkg_str_dup

Reply:
thanks, have incorporated it